### DISCUSSION TITLE
Parallelization based on bootstrap ID

### DISCUSSION THREAD

**Discussion by MahmoudAliAfifi** [nlmixr2/nlmixr2#70] [Q&A]
I am making bootstrap and then fitting each bootstrap dataset parallel, 
please can anyone tell me how to do parallelized fitting based on the ID of the bootstrap dataset?
Also somebody told me that it is better to do parallelization on the level of observation or the level of the bootstrap dataset, I do not know. I was using another R package for PMX and it was taking about 16 hours to finish 500 boostrap datasets.

I am using this code but it does not work
fit <- nlmixr2(one.compartment, bootdatadf, est="saem", control = saemControl(cores=5))

**Comment by mattfidler** [comment]
Hi @MahmoudAliAfifi,

I'm not sure what you are looking at, but `saem` per ID is automatically parallelized in most cases, with the following exceptions (which may change in the future):

- `linCmt()` only models 

- Modeled `rate()` or `dur()` models.

This will occur if you have `OpenMP` enabled.  By default on a mac this is not enabled, though you can enable this yourself and compile from source.  You will immediately know how many cores are used per simulation by loading `rxode2`:

```r
> library(rxode2)
rxode2 2.0.11 using 4 threads (see ?getRxThreads)
```

Which shows I have 4 physical cores on this machine. Typically each machine has 2 threads per processor, so you can be a bit more aggressive by using:

```r
setRxThreads(7)
```

Leaving only 1 thread available for everything else.  Of course you can also set the parallel solving options for `rxode2` in `saemConrtol` by:

```r
control=saemControl(rxControl=rxControl(cores=7))
```

This controls the ode solving (in parallel).  This is a bit different from the original `nlmixr`.  It makes all ODE solving options consistent between each estimation method (all contianed in `rxControl`. 

If you are trying to parallelize the bootstrap, this ma be counterproductive since it runs a `R` parallel process.  Often on linux and maybe on `windows` this is done by `forking` the process.  When forking occurs each worker only uses one thread. Hence, counter to what you may think, parallelization of the bootstrap process for a `saem` model may actually take more time than simply doing each by itself.

One last note, we have a function `bootstrapFit()` that does bootstrapping for you.  It currently isn't parallelized, but to me, this is OK since each estimation problem itself is parallelized. (Note this is not true for `focei`, which still needs to be parallelized)

Hopefully that is enough information.

Note that I will transfer this to a discussion, since it is more of a question/answer.
